### DISCUSSION TITLE
Digital noise on the output

### DISCUSSION THREAD

**Discussion by Hububalli** [ElectroTechnique/electrotechnique.github.io#2] [Software Issues]
Great to see this board available. Hopefully we see lots of input as more people get one of these great little synths.

I did the 100 ohm resistor mod which did improve the noise on the output. I then tried using a wire instead of the trace, pin to pin from the display to the teensy, and that made slight additional improvement. 

Messing around a bit more I noticed that the noise goes completely if FX Mix is turn down. Once you are past about 50% you start to hear digital noise. I assume this must be a separate issue, I don't know much about digital systems but is it possible the effect is overloading the teensy when you apply to much?

Thanks for all your work on this project, It has made a great addition to my setup.

**Comment by cdw2000** [comment]
Yes, I hear this noise as well.  Sounds like a scratchy vinyl record.   From my research, I do not believe it is an issue with the Teensy being overloaded.  I believe it is an issue with the code for the effect_ensemble.cpp, which was written by a 3rd party and added to the Teensy audio library.  Need someone with coding/DSP skills to debug I think (that's not me, unfortunately).

**Reply by Hububalli** [reply]
That is interesting to know. Glad to hear it is likely a software issue. Hopefully someone with the knowledge will be able to shed some light on the problem.

**Reply by seomradubh** [reply]
I also have a scratchy noise on audio over usb to ableton. This does not appear when I plug headphones in to the teensy audio board though??

I did notice that the midi volume is set to loud(est). Dont think this would introduce noise, but i just dont know enough (or anything) about midi to even start troubleshooting that

Does anyone have a definitive answer to this? Do I need to upload some audio examples of this noise?

Thanks

**Comment by Inductor-Hub** [comment]
I have the same observations.
In addition to that, I have a periodig loud noise on the digital USB sound output lasting for 10 seconds every 2 or 3 minutes.
Anyone else ?

**Reply by Electrotechnique** [reply]
No idea what this is. Can you post a sound file?

**Reply by Inductor-Hub** [reply]
Hi ElectroTechnique, hi friends,
thanks for your response. Attached a zipped mp3 file. Its a 20 sec recording of the USB sound output (in Logic). No Mastering no external effects. The Tsynt FX was set to 0%. After sec 1 the noise appears, after 10 sec it disappears slowly. This noise is produced by the Tsynt irregularly after 30-150 sec again. In average, it pops up every 120 seconds.

Nevertheless, Tsynth is a great synth and I love it!

[Tsynth_USB_Noise.zip](https://github.com/ElectroTechnique/electrotechnique.github.io/files/5854422/Tsynth_USB_Noise.zip)

**Reply by Electrotechnique** [reply]
Ok, how are you playing this? Is it via MIDI and what interface?

**Reply by Inductor-Hub** [reply]
I do it via USB midi and the analog audio out is ok

**Reply by Electrotechnique** [reply]
What are you using to record via digital audio? I think that's where the problem is, it's not TSynth if analogue audio is fine.

**Reply by Inductor-Hub** [reply]
Thats good news. I use a Behringer  U-Phoria UMC404HD Audio-Interface for sound processing. I use it for the analg out of the synths to be inserted in Logic and as a soundcard output from Logic. Perhaps the Tsynth and the UMC404 dont like each other. I try with anoter USB interface and let you know.

**Reply by Electrotechnique** [reply]
Check the buffer values in Logic, they probably need to be raised .

**Comment by rolfdegen** [comment]
Hallo friends..
I also build a Teensy 4.1 synthesizer and use the Teensy Audio Lib. I also noticed the disturbance in the sound. It's only with effects and high oscillator levels. The sound without the effect is free of interference.

My video: https://youtu.be/deOFKt0GxbE
My Project: https://forum.pjrc.com/threads/63255-New-Teensy-4-1-DIY-Synthesizer

Greetings from germany. Rolf

**Reply by Inductor-Hub** [reply]
Hi Rolf,
greetings from Munich. I like your synth-project a lot. Do you share the libraries with the Tsynth? I noticed that the sound examples in your video "Jeannie" 8 Voices DIY Synthesizer" are the same as in Tsynth , at least the names and order
. However, they sound of course different.

**Comment by Electrotechnique** [comment]
Hello, the stereo chorus effect produces some noise with a low LFO setting on the FX Amt control. The effect needs either a re-write or replacing with something else. I liked the effect and it uses little CPU.

**Comment by rolfdegen** [comment]
Yes, that's OK. For better effects, I want to add a Spin Semiconductor FV-1 chip to my synthesizer. I built one into my 1st DIY synthesizer. The sound is great Link: https://youtu.be/unN71yPqREU

**Comment by antoniograzioli** [comment]
I think it's a "standard" ground loop noise problem.
if you try to power it up using a mobile phonen PSu, does it still have this noise?
I have noie on mine when connecvted to my mac, it's solved using one of those cheap "groud loop" isolators like: https://it.aliexpress.com/item/32861433401.html

**Reply by Electrotechnique** [reply]
Ground loops are very obvious low hums that go up and down in volume as you touch the connectors.

**Reply by rolfdegen** [reply]
That is a good idea. For my development I use USB isolator e.g. https://www.meilhaus.de/me-usb-iso.htm

**Reply by cdw2000** [reply]
I think these noise topics should be split into separate threads for clarity.   There are two distinct noise issues being discussed here.  One is the digital scratchy noise that comes from use of the Chorus Ensemble effect.   The other is the apparent digital noise over USB audio.

The noise in the Chorus Effect may be solvable with a software modifaction to the chorus code.   The other may be inherent to the design and USB audio libraries for the Teensy.  I don't know.  I haven't noticed the USB noise.

If users hear noise on the output, try turning the Effect Mix knob full counterclockwise to turn the Chorus effect off.  If the noise goes away, then the USB audio is not the source.

**Reply by Electrotechnique** [reply]
If there's noise from the USB digital audio, does increasing the buffer size in for instance, Ableton Live fix it?

**Comment by vipear** [comment]
The scratchy noise is apparently due to quantization of the lfo table values and delay buffer time delays.
I made a branch on my fork in which I modified effect_ensemble files to include a larger lfo table with float values and interpolation between the delay buffer samples. The noise seems to have been substantially improved.

**Reply by Electrotechnique** [reply]
Yes this sounds much better thanks!. I'll put out V2.12 soon.

**Reply by vipear** [reply]
Great! There are two options set with a define. The default uses a larger lfo table that uses about 17KB of flash memory. The second uses a table 1/4 of the memory and uses interpolation to get an equivalent  size. They both sound the same to me. The first one is simpler and takes fewer processor cycles.

**Reply by Electrotechnique** [reply]
Yes, I'll leave it as default. I'll push this to the Teensy3.6 version of TSynth too.

**Reply by vipear** [reply]
Could you use the latest push to my branch? I removed unused code and variables and cleaned up the variable names and scopes so it would be a little less confusing.

**Reply by vipear** [reply]
I didn't update the hex file.

**Reply by Electrotechnique** [reply]
Ok, I've released V2.12 firmware and included it in the previous Teensy3.6 firmware (V1.29). Again, thanks very much!

**Reply by cdw2000** [reply]
Great job on fixing the noise issue in the ensemble effect!   I had been working on my own version but you beat me to it!  I'm fine with that, BTW, since I am not really a C++ programmer of any proficiency.  I did find what I think are cut and paste errors in the code section that does the summation for the various phase taps (lines 218 - 234) .   For the 4th and 5th phases, you are adding outIndex3 to offset4 and offset5, where I think outIndex4 and outIndex5 should be used instead.

Also, the values of the lfoIndexes need to be corrected.  These have been wrong from the beginning with the original 3 phase effect_ensemble.cpp file.   Given that one period of the modulation waveform is represented in 1470 samples, the indexes for 120 and 240 degrees should be 490 and 980, respectively, but the original code used half these values.   Also to get 60 degree spacing, you need 6 indexes, not 5, so I added the 6th index and corrected all the values.   I think it sounds pretty smooth now,

These fixes/changes are in the Tsynth 4.1 fork I made, if you want to try them out.

**Reply by cdw2000** [reply]
I have also been doing a lot of research on the original analog circuit on which this ensemble effect is based.   Alexander Davis assumed the two waveforms were at 6 Hz and 0.6 Hz, or a 10:1 ratio.  In reality the analog circuits were at 0.66 Hz and 6.25 Hz which is closer to a 9:1 ratio.   Also the amplitude of the "vibrato" 6 Hz waveform was not as high as what is being used in Alexander's wave formula.   I found these mods are closer to the original analog circuit:

y=((sin(((2.0 * pi)/LFO_SAMPLES) * x) * LFO_RANGE) / 2.0) + (((sin(((**18.0** * pi)/LFO_SAMPLES) * x)) * LFO_RANGE) / **5.0**)

@vipear, would you be willing to share your Octave script for generating the lfoTable?

**Reply by vipear** [reply]
Your fix sounds great! Thanks for pointing out and fixing the copy/paste error. I was wondering why the taps went from 3 to 5 and not 6. I was trying not to second guess the top level algorithm from the original author as I probably have less experience in how it would sound. I was more focused on the implementation that caused the noise. I'm an OK programmer, not an expert; my specialty is in analog electronics. I have never used DMA before so the TSynth code is a learning experience.
I have pushed the Octave script lfo_script.m to my branch in the source code directory.
I was trying to figure out the relationship of the table and LFO frequency but I couldn't quite figure it out.
From what I see, the variable countsPerLfo is set to 200 in the original code when the control is set to 6Hz. That means that for every 200 audio samples (I assume 44.1KHz) the lfo index increments by one, or at a 220.5 Hz rate. Getting through the 1470 table samples would happen at a 220.5/1470 = 0.15Hz rate. Is that right or am I missing something?

**Reply by cdw2000** [reply]
Yes, I also believe the base modulation frequency at the original 200 counts equals 0.15 Hz.    I did some listening tests by removing all but the single 0 phase tap so I could gauge the modulation frequency by ear and 0.15 Hz sounds about right.   From Simon's (@ElectroTechnique) comments - when he modified the code he assumed 200 counts = 6 Hz.   From the original coder's comments either 6 Hz or 0,6 Hz was a reasonable assumption, but it's not really the case.   I made a version of the TSynth that displays the correct frequency and provides a slightly wider control span, but it would mean all of the presets would have to be modified to correct for the new frequency settings.   Interestingly 0.6Hz and 6Hz sounds a little too fast for me.   The current TSynth range sounds good, so I chose not raise the frequency as an issue.   BTW, I also am more of an analog circuit guy.   I figured out the correct frequencies and waveforms by using LTSpice to model the actual Roland RS-202 ensemble circuit.

**Reply by vipear** [reply]
I tried out your suggested 9:1 ratio formula and without being able to do a quick A/B comparison I would say my feeble ears detected a little more "motion" in the sound.

**Reply by Electrotechnique** [reply]
Ok, I'm testing it and it sounds great. I'm getting clicking on 3 - Overhead when playing lots of notes quite low down and I think this is volume overload, so I'll drop the voice gain a bit which seems to stop it.

Can we come up with an agreement on what the range of FX Amt is? At the moment it's 2Hz-20Hz, which I know isn't correct. I'll consider how to display this and change the presets if necessary.

**Reply by cdw2000** [reply]
I don't think you need to adjust the voice gain.   In the original effect_ensemble, with 3 taps (phases), there is a divide by 3 term, since you are effectively summing 3 waveforms together.   When the number of waveforms was increased, the divisor wasn't changed.  I just updated my fork to change the divisor to 6 to fix that.  See if that resolves the overload.    Since the waveforms aren't coherent, we might be able to get away with a lower divisor of 5 or 4 without overload.   Note that my fork doesn't yet modify the waveform for a 9:1 ratio.

**Reply by Electrotechnique** [reply]
Yes, that's it, it's fine now.

**Reply by cdw2000** [reply]
Regarding the fxamt display -.   I worked out that the current display range of 2 to 20 Hz results in an actual LFO frequency of 0.05 to 0.5 Hz.   Which is interesting because it sounds pretty fast, but doesn't get to the 0.6 Hz of the original analog circuit.   I propose a range of 0.05 Hz to 1.00 Hz.   The countsPerLfo formula should be:

	countsPerLfo = round(AUDIO_SAMPLE_RATE_EXACT / (LFO_SAMPLES * rate));
	if(countsPerLfo < 1) {
		countsPerLfo = 1;
	}

I have a new constants table for the fxamt values, as well, but I have to go to my day job now.   I will update about 10 hours from now...

**Reply by Electrotechnique** [reply]
Ok, I'll leave this until I'm told what to do with it! It does sound considerably better.

**Reply by cdw2000** [reply]
I compared the 10:1 ratio LFO waveform to the 9:1 ratio version, and I think I prefer the original 10:1 version.   Maybe it has to do with using 6 phases versus the original 3.   I opted not to merge my branch with the 9:1 ratio, so my fork has the bug fixes but still uses the 10:1 ratio.    I'm thinking maybe this should be released as a minor bug fix to V2.12 before looking to update the LFO frequency display changes, since that will be a bit disruptive.

**Comment by Electrotechnique** [comment]
**If you'd like to make a Pull Request to merge any changes from now on, please do so. I'm going to start adding people as collaborators, so that I'm no longer the _bottleneck_ in future development**

**Reply by cdw2000** [reply]
I created 2 pull requests for the fixes to the effect_ensemble.cpp and effect_ensemble.h files.

**Comment by glido1** [comment]
The electronic design still suffers from several childhood diseases.
How to solve noise when the Tsynth is powered from the computer's USB?
A galvanic isolator is needed.
USB Isolator ADUM4160 ADUM3160.
It costs $ 10 and solves the problem.
The best solution would be to place a DC / DC converter, e.g. Ricoh, on the PCB, like for example Yamaha.
You can also opt out of TSynth power supply via USB Host and provide power, for example, from the charger for the HTC10 smartphone, which have a noise reduction regulator, and there will be silence, without interference.

Regarding the TSYNTH itself, it needs a correction on the analog output.
Soon I will publish the recording on YT.
TSYNTH + DAC Es9023 + Tube Ecc83 Hot EQ with Volume control.
The SGTL5000 is a weak DAC.Ground loops should be avoided.
The analog ground should connect to the DGND with a jumper and not a thick rail, If the SGTL 5000 has such a poor low-pass filter, you need to build a second opamp, e.g. 22kHz. Capacitors and a 12V stabilizer for an opamp, e.g. opa 2604, will filter out the noise from the DGND ground. 
Regards Robert

Google translator - sorry for Eanglish.
I only know German

**Comment by glido1** [comment]
https://www.fotosik.pl/zdjecie/63277ff7427342c4
Everything here is to get rid of digital noise on the output
